R/adducts.R

```diff
@@ -127,6 +180,108 @@ mz2mass <- function(x, adduct = "[M+H]+") {
                 "data.frame with the adduct definition")
 }
 
+#' @title Calculate a table of adduct (ionic) formulas
+#' 
+#' @description `adductFormula` calculates the adduct formulas corresponding
+#'   from the combinations of a list of formulas and a list of adducts.
+#'
+#' @param formulas `character` list of molecular formulas.
+#'
+#' @param adducts `character` list or `data.frame` of valid adducts to be used.
+#'   Custom adduct definitions can be provided via a `data.frame` but its format
+#'   must follow [adducts()]
+#'
+#' @param standardize `logical` whether to standardize the molecular formulas to
+#'   the Hill notation system before calculating their mass.
+#'
+#' @return `character` matrix with *formula* rows and *adducts* columns
+#'   containing all ion formulas. In case an ion can't be generated (eg.
+#'   \[M-NH3+H\]+ in a molecule that doesn't have nitrogen), a NA is returned
+#'   instead.
+#'
+#' @seealso [adductNames()] for a list of all available predefined adducts and
+#'   [adducts()] for the adduct `data.frame` definition style.
+#' 
+#' @author Roger Gine
+#' 
+#' @examples 
+#' 
+#' # Calculate the ion formulas of glucose with adducts [M+H]+, [M+Na]+ and [M+K]+
+#' adductFormula("C6H12O6", c("[M+H]+", "[M+Na]+", "[M+K]+"))
+#' 
+#' # > "[C6H13O6]+" "[C6H12O6Na]+" "[C6H12O6K]+"
+#' 
+#' # Use a custom set of adduct definitions (For instance, a iron (Fe2+) adduct)
+#' custom_ads <- data.frame(name = "[M+Fe]2+", mass_multi = 0.5, charge = 2,
+#'                          formula_add = "Fe", formula_sub = "C0",
+#'                          positive = "TRUE")
+#' adductFormula("C6H12O6", custom_ads)
+#' 
```

Comment:
I tested some additional formulas:

this returns `NA` because it simply doesn't work, which I think is OK

```r
> adductFormula("C6H2O1", adducts = "[M+H-NH3]+")
      [M+H-NH3]+
C6H2O         NA
```

but for this one I think it would be better if elements with a `0` count  would be dropped?

```r
> adductFormula("C6H2O1N1", adducts = "[M+H-NH3]+")
       [M+H-NH3]+  
C6H2NO "[C6H0N0O]+"
```

Reply:
yes, I think this would definitely be helpful

Reply:
Sure thing!  :+1: